### PR TITLE
Check link

### PR DESCRIPTION
**Pull request by HugoTian**
check before link and unlink to see whether has write permission
check before download and extract to see whether has enough space

### CODE REVIEW

**Comment by HugoTian** [comment]
@kalefranz , new PR for check permission and disk space

**Comment by kalefranz** [comment]
times look good now

File: conda/fetch.py [hunk 1]
```diff
@@ -25,9 +25,13 @@
 from .exceptions import (ProxyError, ChannelNotAllowed, CondaRuntimeError, CondaSignatureError,
                          CondaHTTPError)
 from .install import (add_cached_package, find_new_location, package_cache, dist2pair,
-                      rm_rf, exp_backoff_fn)
+                      rm_rf, exp_backoff_fn, on_win)
 from .lock import Locked as Locked
 from .utils import memoized
+from .exceptions import ProxyError, ChannelNotAllowed, CondaRuntimeError, \
+    CondaSignatureError, CondaIOError
```
Comment (kalefranz) [inline conda/fetch.py]: while it's just style, here I strongly prefer

    from .exceptions import (ProxyError, ChannelNotAllowed, CondaRuntimeError,
                             CondaSignatureError, CondaIOError)

File: conda/instructions.py [hunk 2]
```diff
@@ -132,6 +135,11 @@ def execute_instructions(plan, index=None, verbose=False, _commands=None):
         if cmd is None:
             raise InvalidInstruction(instruction)
 
+        # check before link and unlink package
+        if cmd in [LINK_CMD, UNLINK_CMD] and not checked:
```
Comment (kalefranz) [inline conda/instructions.py]: Use a tuple instead of a list. It's faster.

File: conda/install.py [hunk 3]
```diff
@@ -839,12 +838,21 @@ def extract(dist):
...
         temp_path = path + '.tmp'
         rm_rf(temp_path)
+
+        # check whether has enough space to extract a file
+        from .fetch import check_size
         with tarfile.open(fname) as t:
+            size = 0
+            for m in t.getmembers():
+                size += m.size
+            check_size(pkgs_dir, size)
```
Comment (kalefranz) [inline conda/install.py]: I think this is the wrong place to do this.  I think we should sum the sizes of all the packages we're about to download and extract, and then do just one check.

File: conda/lock.py [hunk 4]
```diff
@@ -49,6 +49,15 @@ def touch(file_name, times=None):
         os.utime(file_name, times)
 
 
+def preprocess_name(path):
```
Comment (kalefranz) [inline conda/lock.py]: I think Locked (or the FileLock it turned into), might need another assert.  `assert not path.startswith('http')`

We also have utility functions to convert a `file://` url to a path.  I think it's `util.url_to_path`.  There's also a `connection.path_to_url` I think.  Someday we'll clean up where all this stuff is.

File: tests/test_fetch.py [hunk 5]
```diff
@@ -28,9 +29,16 @@ def test_fetchrepodate_connectionerror(self):
             fetch_repodata(url)
             assert msg in str(execinfo)
 
+    def test_fetchrepodate_connectionerror(self):
+        with pytest.raises(RuntimeError) as execinfo:
+            url = "http://10.0.0.0/"
```
Comment (kalefranz) [inline tests/test_fetch.py]: We had changed this to`240.0.0.0`.  I think the PR is probably old (my fault) and you should rebase.

**Comment by kalefranz** [comment]
I didn't see anything about checking for write permissions here.  I guess this is still WIP?

File: conda/instructions.py (hunk 2 above)
Comment (HugoTian) [inline conda/instructions.py]: Great, happy to learn that

File: conda/install.py (hunk 3 above)
Comment (HugoTian) [inline conda/install.py]: I agree ! check in advance will be good, but there is complexity in implementation. In fetch and extraction, we get the path of file in extract and fetch method, so it is simple to check there. However, if we check in advance, we need to figure out the destination first, and they may go to different places.

Or we can just check whether there is enough space in prefix, this is a simple solution

**Comment by HugoTian** [comment]
For the checking place, I agree with you idea ! 

Check in advance will be good, but there is complexity in implementation. In fetch and extraction, we get the path of file in extract and fetch method, so it is simple to check there. However, if we check in advance, we need to figure out the destination first, and they may go to different places.

Or we can just check whether there is enough space in prefix, this is a simple solution

**Comment by kalefranz** [comment]
Let's walk through the fetch code together tomorrow and see if we can make it less complex to implement.

**Comment by kalefranz** [comment]
CC @dorzel please review. 

Would like to try to get this in the next canary release.

**Comment by kalefranz** [comment]
Superseded by #3301

**Pull request closed without merging**